tests/impl.cpp

```diff
@@ -3095,6 +3095,17 @@ result_t test_mm_extract_epi16(const SSE2NEONTestImpl &impl, uint32_t i)
     return TEST_UNIMPL;
 }
 
+result_t test_mm_extract_pi16(const SSE2NEONTestImpl &impl, uint32_t i)
+{
+    uint16_t *_a = (uint16_t *) impl.mTestIntPointer1;
+    const int imm = 1;
+
+    const __m64 *a = ((const __m64 *) _a);
+    int c = _mm_extract_pi16(*a, imm);
+    ASSERT_RETURN((uint16_t) c == *(_a + imm));
```

Comment:
The casting should be `uint16_t` to `int`.
The return value of `_mm_extract_pi16` is `int`, and we should focus on it.

Be aware that the test of `mm_extract_epi16` might have the same problem.

Reply:
But `_mm_extract_pi16` takes arithmetic shift operation, which means expected value 65535 will return -1 by `_mm_extract_pi16`

Reply:
Then I think the comparison should include two parts, which are `c[31:16]` and `c[15:0]`.

Reply:
sorry I don't understand what you expect

Reply:
```
dst[15:0] := (a[63:0] >> (imm8[1:0] * 16))[15:0]
dst[31:16] := 0
```
The test should verify this.

Reply:
I found the behavior is not actually the same as the pseudocode you showed
with `make check`  and the following test I got such response

```
int16_t *_a = (int16_t *) impl.mTestIntPointer1;
    const int imm = 1;

    const __m64 *a = ((const __m64 *) _a);
    int c = _mm_extract_pi16(*a, imm);

    printf("_a= %d, %d, %d, %d\n", _a[0], _a[1], _a[2], _a[3]);
    printf("c= %d, (c & 0xFFFF0000) = %d\n",c,  (c & 0xFFFF0000));
```

this will generate this response 
```
_a= 13943, -1, -28555, 0
c= -1, (c & 0xFFFF0000) = -65536
```

this response shows intrinsics, `_mm_extract_pi16` doesn't set the leading 16 bits as zeros, instead it just takes one of the  16 bits integer from a 64 bits long integer

Reply:
so keeping the current test may be a better idea I think

Reply:
The reason of the mismatching behaviour should be found.
I wonder whether this is caused by the compiler or not.

I will try to investigate.

Reply:
The document of [PEXTRW - Extract Word](https://www.felixcloutier.com/x86/pextrw) says that the upper bits should be set to 0.
And the reported [bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98495) of GCC Bugzilla says that `X86 _mm_extract_pi16 incorrectly sign extends result`.

Please modify the test as I suggested and implement the test for `Clang` only.
The `GCC` part should be left unimplemented and add the corresponding comment that indicates it is caused by a bug and might be fixed in `GCC-11`.